Failed to get log when GC triggered by cron

What can we help you?

After having an investigation, it should be an issue of jobservice, the cron job will generate a log with name like `6ca39099ba856921ee1fc4d0@1585035359`. When core tries to use the name to get log content, jobsservice throws an error: 

`Mar 24 07:47:59 172.30.0.1 jobservice[19235]: 2020-03-24T07:47:59Z [ERROR] [/jobservice/api/handler.go:288]: Serve http request 'GET /api/v1/jobs/6ca39099ba856921ee1fc4d0@1585035359/log' error: 500 {"code":10009,"message":"failed to get the job log","details":"invalid length of log identify"}`

--- a/src/jobservice/logger/getter/file_getter.go
+++ b/src/jobservice/logger/getter/file_getter.go
@@ -6,6 +6,7 @@ import (
 	"fmt"
 	"io/ioutil"
 	"path"
+	"strings"
 
 	"github.com/goharbor/harbor/src/jobservice/errs"
 
@@ -24,12 +25,8 @@ func NewFileGetter(baseDir string) *FileGetter {
 
 // Retrieve implements @Interface.Retrieve
 func (fg *FileGetter) Retrieve(logID string) ([]byte, error) {
-	if len(logID) != 24 {
-		return nil, errors.New("invalid length of log identify")
-	}
-
-	if _, err := hex.DecodeString(logID); err != nil {
-		return nil, errors.New("invalid log identify")
+	if err := isValidLogID(logID); err != nil {
+		return nil, err
 	}
 
 	fPath := path.Join(fg.baseDir, fmt.Sprintf("%s.log", logID))
@@ -39,4 +36,22 @@ func (fg *FileGetter) Retrieve(logID string) ([]byte, error) {
 	}
 
 	return ioutil.ReadFile(fPath)
+}
+
+func isValidLogID(id string) error {
+	lid := id
+	segment := strings.LastIndex(lid, "@")
+	if segment != -1 {
+		lid = lid[:segment]
+	}
+
+	if len(lid) != 24 {
+		return errors.New("invalid length of log identify")
+	}
+
+	if _, err := hex.DecodeString(lid); err != nil {
+		return errors.New("invalid log identify")
+	}
+
+	return nil
 }

